rand_core/src/lib.rs

```diff
@@ -247,55 +247,91 @@ pub enum ErrorKind {
     __Nonexhaustive,
 }
 
-#[cfg(feature="std")]
+impl ErrorKind {
+    pub fn description(&self) -> &'static str {
+        match *self {
+            ErrorKind::Unavailable => "permanent failure or unavailable",
+            ErrorKind::Transient => "transient failure",
+            ErrorKind::NotReady => "not ready yet",
+            ErrorKind::Other => "uncategorised error",
+            ErrorKind::__Nonexhaustive => unreachable!(),
+        }
+    }
+}
+
 #[derive(Debug)]
-pub struct Error {
-    pub kind: ErrorKind,
-    pub cause: Option<Box<std::error::Error>>,
+enum ErrorDetails {
+    None,
+    Str(&'static str),
+    #[cfg(feature="std")]
+    Error(Box<std::error::Error + Send + Sync>),
 }
 
-#[cfg(not(feature="std"))]
-#[derive(Debug, Clone, PartialEq, Eq)]
+/// Error type of random number generators
+/// 
+/// This embeds two things, an `ErrorKind`, which can be matched over, and
+/// optional details provided to `new_str` or `new_err`.
+#[derive(Debug)]
 pub struct Error {
+    /// Error kind
     pub kind: ErrorKind,
-    pub cause: Option<&'static str>,
+    details: ErrorDetails,
 }
 
 impl Error {
+    /// Create a new instance, with no details.
+    pub fn new(kind: ErrorKind) -> Self {
+        Self { kind, details: ErrorDetails::None }
+    }
+    
+    /// Create a new instance, with a static string describing the details.
+    /// 
+    /// This may be used in both `std` and `no_std` enviroments.
+    pub fn new_str(kind: ErrorKind, details: &'static str) -> Self {
+        Self { kind, details: ErrorDetails::Str(details) }
+    }
+    
+    /// Create a new instance, with a chained `Error` cause.
     #[cfg(feature="std")]
-    pub fn new(kind: ErrorKind, cause: Option<Box<std::error::Error>>) -> Error {
-        Error {
-            kind: kind,
-            cause: cause,
+    pub fn new_err<E>(kind: ErrorKind, cause: E) -> Self
```

Comment:
Does it make sense to have a function that takes both a kind and a string, and use one of them depending on whether the `std` feature is enabled? Now implementations of `Rng` need a `std` feature flag to give the right type of cause.

Reply:
I don't get what you mean? This is more generic than `new_str` and can still accept a string but is only available with the `std` library (we could probably depend on `alloc` instead, but that's another matter; we need `Box`).

Reply:
See [this comment](https://github.com/dhardy/rand/issues/10#issuecomment-338666735).

Reply:
The `Into` where clause it neat, wouldn't have thought about that myself.  Although maybe it is common, and I am just missing something :-).

I was thinking about cases like [this](https://github.com/dhardy/rand/pull/28/files#diff-412db97196316ae7812d61b5624f0ad9R162), where a `Rng` implementation needs to use different functions to construct an error depending on whether the `std` feature is available.

Then a helper function is maybe useful:
``` rust
The `Into` where clause it neat, wouldn't have thought about that myself.  Although maybe it is common, and I am just missing something :-).

``` rust
#[cfg(feature="std")]
pub fn new_with_cause(kind: ErrorKind, cause: E, details: &'static str) -> Self {
    new_str(kind, details)
}

#[cfg(not(feature="std"))]
pub fn new_with_cause(kind: ErrorKind, cause: E, details: &'static str) -> Self {
    new_err(kind, cause)
}
```

Reply:
In `no_std` environment you simply can't have implementation of `Error` trait, as it defined only in the `std`. So either way crate authors will have to write `cfg`s to handle switching.

Reply:
Or authors supporting `no_std` can simply use `new_str` with a string description everywhere. There's no advantage to using `new_err` if the details are just a string.

Reply:
I updated my branch with `JitterRng` to the new error handling https://github.com/pitdicker/rand/commit/b223c60539680490c98316fd1d4f8b3dcb8b17ca.

The second commit shows what I mean https://github.com/pitdicker/rand/commit/f6b353154b8dc8083ddb78f77ad1f91634098b8e.
In theory an RNG implementation can support both `std` and `no_std` compatible error causes, and the choice is made in `rand_core`.

Reply:
It was just an idea, but I don't think it helps in enough cases. Lets leave the two functions as is.

Reply:
Your `new_with_cause` isn't even going to compile in `no_std` mode since it depends on `std::error::Error`.

I see what you mean, but I'm not convinced it's worth returning a chained `Error` type in this case; there is no added information (but maybe your `description()` strings should include all the details from the `Display` impl).

Reply:
Oops. Ok, please forget I said anything :-)